{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "72712eb1",
   "metadata": {},
   "source": [
    "# Multiple Co-Inertia Analysis (MCoIA) Documentation\n",
    "\n",
    "The MCoIA package is a versatile Python toolkit designed for advanced\n",
    "data analysis, particularly suited for multi-omics datasets. It excels\n",
    "at integrating data from various sources and modalities, such as\n",
    "genomics, proteomics, and metabolomics, to uncover patterns and\n",
    "relationships that are often obscured in single-dataset analyses. By\n",
    "leveraging the principles of co-inertia analysis, MCoIA facilitates a\n",
    "deeper understanding of complex biological systems, aiding researchers\n",
    "in fields ranging from systems biology to personalized medicine. This\n",
    "package draws inspiration from the ADE4 package in R. This documentation\n",
    "provides a comprehensive guide on installation, usage, and development\n",
    "of MCoIA.\n",
    "\n",
    "## Getting Started\n",
    "\n",
    "You can download this package from <https://github.com/Akarr0x/mcioa>\n",
    "\n",
    "To install the package, run the following commands in your terminal:\n",
    "\n",
    "``` bash\n",
    "git clone https://github.com/Akarr0x/mcioa.git\n",
    "cd mcioa\n",
    "python -m pip install .\n",
    "```\n",
    "\n",
    "## User Guide\n",
    "\n",
    "Explore detailed documentation on how to use MCoIA, covering all its\n",
    "features and functionalities.\n",
    "\n",
    "-   This is the main use of the MCoIA package:\n",
    "\n",
    "``` python\n",
    "import mcoia\n",
    "\n",
    "data_list = [dataset1, dataset2]\n",
    "\n",
    "mcia_instance = MCIAnalysis(data_list)\n",
    "```\n",
    "\n",
    "`MCIAnalysis()` This function initializes the object that will be used\n",
    "for further analysis. Its primary purpose is to set up and initialize\n",
    "the necessary variables. The `MCIAnalysis()` method requires a list of\n",
    "pandas DataFrame objects as its input. Each DataFrame should contain\n",
    "structured data corresponding to a particular block of the multi-block\n",
    "dataset you wish to analyze. Ensure that the DataFrames are\n",
    "pre-processed and aligned such that rows correspond to samples and\n",
    "columns to features for consistent analysis across blocks.\n",
    "\n",
    "``` python\n",
    "mcia_instance.fit()\n",
    "```\n",
    "\n",
    "`fit()` This function prepares the data for the core MCoIA computations\n",
    "by normalizing each dataset independently. The default normalization\n",
    "method is Non-Symmetric Correspondence Analysis. However, this can be\n",
    "altered using the `analysis_type` parameter. By setting `analysis_type`\n",
    "to 'pca', a different normalization is applied where each cell is\n",
    "weighted equally.\n",
    "\n",
    "``` python\n",
    "mcia_instance.transform()\n",
    "```\n",
    "\n",
    "`transform()` This function introduces necessary block differentiation\n",
    "within the datasets. It is essential for allocating distinct weights\n",
    "across the datasets appropriately and facilitates the computation of row\n",
    "and column coordinates in the subsequent analysis steps.\n",
    "\n",
    "``` python\n",
    "mcia_instance.results()\n",
    "```\n",
    "\n",
    "`results()` This function executes the actual multiple co-inertia\n",
    "calculations as described in <a href=\"#Dray2007\" class=\"citation\">[Dray2007]</a>. It\n",
    "enables the visualization of the dataset in a reduced dimensional space,\n",
    "optimizing for maximum covariance between the datasets.\n",
    "\n",
    "The variables `dataset1` and `dataset2` are expected to be pandas\n",
    "DataFrame objects. The output will include the following variables:\n",
    "\n",
    "-   `pseudo_eigenvalues`: Eigenvalues derived from the singular value\n",
    "    decomposition (SVD) on the concatenated datasets, representing the\n",
    "    variance captured by each principal component.\n",
    "\n",
    "-   `lambda_df`: A matrix where each element (i, j) corresponds to the\n",
    "    eigenvalue associated with the ith block's jth factor, summarizing\n",
    "    the eigenvalues across blocks and factors.\n",
    "\n",
    "-   `SynVar`: Contains the left singular vectors from the SVD,\n",
    "    representing the synthetic variables that capture the most variance\n",
    "    within the data.\n",
    "\n",
    "-   `axis`: Comprises the right singular vectors from the SVD,\n",
    "    indicating the principal axes in the factor space.\n",
    "\n",
    "-   `row_projection`: The projection of the original dataset's rows\n",
    "    onto the principal components, providing a weighted representation\n",
    "    in the reduced factor space.\n",
    "\n",
    "-   `cov2`: A scalar value representing the total covariance explained\n",
    "    by the current factors, offering a measure of the explanatory power\n",
    "    of the analysis.\n",
    "\n",
    "-   `row_projection_normed`: Similar to `row_projection`, but normalized\n",
    "    to unit length, presenting a standardized view of the dataset's row\n",
    "    projections.\n",
    "\n",
    "-   `column_projection`: The projection of the dataset's columns,\n",
    "    providing a weighted representation of the variables in the factor\n",
    "    space.\n",
    "\n",
    "-   `class_type`: Indicates the classification or grouping assigned to\n",
    "    each object within the analysis, useful for downstream\n",
    "    categorization or stratification.\n",
    "\n",
    "    > -   It is also possible to invoke the `project()` method with a\n",
    "    >     different dataset to project it onto the space defined by the\n",
    "    >     MCoIA calculation. This can be achieved by calling\n",
    "    >     `MCIAnalysis.project(dataset3)`.\n",
    "\n",
    "## Using MCoIA with Example Datasets\n",
    "\n",
    "To illustrate how to use the MCoIA package, consider the following\n",
    "example where we have two datasets represented as lists of values. These\n",
    "lists are then converted into pandas DataFrame objects with\n",
    "corresponding gene names and cell names as indices and columns,\n",
    "respectively.\n",
    "\n",
    "``` python\n",
    "dataset1_values = [\n",
    "    [23, 43, 65, 22, 1, 78],\n",
    "    [34, 23, 45, 65, 23, 43],\n",
    "    [45, 67, 23, 54, 23, 65],\n",
    "    [56, 43, 23, 43, 23, 54],\n",
    "    [67, 65, 34, 65, 12, 43]\n",
    "]\n",
    "\n",
    "dataset2_values = [\n",
    "    [34, 56, 23, 12, 43, 23],\n",
    "    [45, 34, 56, 54, 23, 54],\n",
    "    [65, 43, 23, 43, 34, 23],\n",
    "    [23, 12, 34, 65, 43, 65],\n",
    "    [43, 23, 65, 34, 23, 54]\n",
    "]\n",
    "\n",
    "gene_names = [f\"Gene_{i}\" for i in range(1, 6)]\n",
    "col_names = [f\"Cell_{i}\" for i in range(1, 7)]\n",
    "```\n",
    "\n",
    "Next, we create pandas DataFrames from these lists and perform the MCoIA\n",
    "analysis:\n",
    "\n",
    "``` python\n",
    "import pandas as pd\n",
    "from mcioa import MCIAnalysis\n",
    "\n",
    "# Create DataFrames\n",
    "\n",
    "data_list = [dataset1, dataset2]\n",
    "\n",
    "mcia_instance = MCIAnalysis(data_list)\n",
    "\n",
    "mcia_instance.fit()\n",
    "\n",
    "mcia_instance.transform()\n",
    "\n",
    "mcia_instance.results()\n",
    "half_size = mcia_instance.column_projection.shape[0] // 2\n",
    "\n",
    "data1 = mcia_instance.column_projection.iloc[:half_size]\n",
    "data2 = mcia_instance.column_projection.iloc[half_size:]\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.scatter(data1['SV1'], data1['SV2'], s=10, c=\"b\", marker=\"s\", label=\"first\")\n",
    "ax.scatter(data2['SV1'], data2['SV2'], s=10, c=\"r\", marker=\"o\", label=\"second\")\n",
    "ax.legend()\n",
    "plt.show()\n",
    "```\n",
    "\n",
    "## Developer Guide\n",
    "\n",
    "Welcome to the MCoIA community! My name is Alessandro, the creator and\n",
    "current sole developer of this package, developed as part of my\n",
    "master's thesis. While I'm continually learning in this field, I'm\n",
    "excited to collaborate with others who are interested in contributing,\n",
    "regardless of their level of expertise. Whether it's improving the\n",
    "code, enhancing documentation, or suggesting new features, every\n",
    "contribution is valuable. For collaboration or contributions, please\n",
    "feel free to use the GitHub issues or discussions for this project.\n",
    "\n",
    "Setting Up Development Environment:\n",
    "\n",
    "``` bash\n",
    "git clone https://github.com/Akarr0x/mcioa.git\n",
    "cd mcioa\n",
    "python -m venv venv\n",
    "source venv/bin/activate  # On Windows use `venv\\Scripts\\activate`\n",
    "pip install -r requirements.txt\n",
    "pip install .\n",
    "\n",
    "# Running tests to verify installation\n",
    "pytest test\n",
    "```\n",
    "\n",
    "---\n",
    "\n",
    "**How to Contribute**\n",
    "\n",
    "Welcome to the MCoIA project! Here's a simple step-by-step guide to get\n",
    "you started:\n",
    "\n",
    "1.  **Forking the Repository**: First, visit the [MCoIA GitHub\n",
    "    repository](https://github.com/Akarr0x/mcioa) and click on the\n",
    "    'Fork' button. This creates a copy of the project in your GitHub\n",
    "    account.\n",
    "2.  **Creating a New Branch**: After forking, clone the repository to\n",
    "    your local machine and create a new branch for your contributions:\n",
    "\n",
    "``` bash\n",
    "git checkout -b my-new-feature\n",
    "```\n",
    "\n",
    "1.  **Making and Committing Changes**: Make your desired changes in this\n",
    "    new branch. Once you're done, commit these changes with a clear\n",
    "    message:\n",
    "\n",
    "``` bash\n",
    "git commit -m \"Add a brief description of your changes\"\n",
    "```\n",
    "\n",
    "1.  **Pushing Changes to Your Fork**: Push your new branch and changes\n",
    "    to your forked repository on GitHub:\n",
    "\n",
    "``` bash\n",
    "git push origin my-new-feature\n",
    "```\n",
    "\n",
    "1.  **Creating a Pull Request**: Go to your fork on GitHub and click\n",
    "    'New Pull Request'. Describe your changes and then submit the pull\n",
    "    request.\n",
    "\n",
    "**Code of Conduct**\n",
    "\n",
    "We highly value respect and kindness. Our community embraces the\n",
    "principles outlined in the Python Software Foundation's Code of\n",
    "Conduct, which can be found [here](https://www.python.org/psf/conduct/).\n",
    "These values represent the core of how we interact and collaborate.\n",
    "\n",
    "**Community Guidelines**\n",
    "\n",
    "We believe in a collaborative and open environment. Here's how you can\n",
    "actively participate:\n",
    "\n",
    "-   **Discussing Features**: Got an idea for a new feature or\n",
    "    improvement? We'd love to hear about it! Please open a 'Feature\n",
    "    Request' issue on our GitHub repository to start the conversation.\n",
    "\n",
    "**Reporting Bugs/Issues**\n",
    "\n",
    "If you encounter a bug, we'd appreciate your help in fixing it:\n",
    "\n",
    "1.  **Check Existing Issues**: Before reporting, please check if the bug\n",
    "    has already been reported. If it has, feel free to add any\n",
    "    additional information you might have to the existing report.\n",
    "2.  **Open a New Issue**: If the bug hasn't been reported yet, open a\n",
    "    new issue on our GitHub repository. Provide as much detail as\n",
    "    possible, including steps to reproduce the bug, and any relevant\n",
    "    screenshots or error messages.\n",
    "\n",
    "Your contributions and suggestions, no matter how big or small, are\n",
    "highly valued.\n",
    "\n",
    "---\n",
    "\n",
    "# References\n",
    "\n",
    "The development of MCoIA was informed by the following works:\n",
    "\n",
    "<span id=\"Bougeard2018\" class=\"citation-label\">Bougeard2018</span>  \n",
    "Bougeard, S., & Dray, S. (2018). \"Supervised Multiblock Analysis in\n",
    "R with the ade4 Package.\" Journal of Statistical Software, 86(1),\n",
    "1-17. <doi:10.18637/jss.v086.i01>.\n",
    "\n",
    "<span id=\"Chessel2004\" class=\"citation-label\">Chessel2004</span>  \n",
    "Chessel, D., Dufour, A., & Thioulouse, J. (2004). \"The ade4 Package\n",
    "– I: One-Table Methods.\" R News, 4(1), 5-10. [Available\n",
    "online](https://cran.r-project.org/doc/Rnews/).\n",
    "\n",
    "<span id=\"Dray2007\" class=\"citation-label\">Dray2007</span>  \n",
    "Dray, S., & Dufour, A. (2007). \"The ade4 Package: Implementing the\n",
    "Duality Diagram for Ecologists.\" Journal of Statistical Software,\n",
    "22(4), 1-20. <doi:10.18637/jss.v022.i04>.\n",
    "\n",
    "<span id=\"Dray2007-2\" class=\"citation-label\">Dray2007-2</span>  \n",
    "Dray, S., Dufour, A., & Chessel, D. (2007). \"The ade4 Package –\n",
    "II: Two-Table and K-Table Methods.\" R News, 7(2), 47-52. [Available\n",
    "online](https://cran.r-project.org/doc/Rnews/).\n",
    "\n",
    "<span id=\"Thioulouse2018\" class=\"citation-label\">Thioulouse2018</span>  \n",
    "Thioulouse, J., Dray, S., Dufour, A., Siberchicot, A., Jombart, T.,\n",
    "& Pavoine, S. (2018). Multivariate Analysis of Ecological Data with\n",
    "ade4. Springer. <doi:10.1007/978-1-4939-8850-1>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da475e0a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
